source/geod24/Serialization.d

```diff
@@ -0,0 +1,259 @@
+    only on what operations should be supported.
+
+    For a type `T` to be a valid serializer, it should support
+    the following actions:
+    - `T.serialize()` should compile when passed on parameter.
+    - This one parameter can be any type that is either a parameter or a
+      return value of `API`'s methods. This is not checked explicitly,
+      but will lead to compilation error in LocalRest if not followed.
+    - `T.serialize(param)` should return an one of `string`,
+      `immutable(void)[]`, `immutable(ubyte)[]`
+    - `T.deserialize!QT` is a template function that returns an
+      instance of type `QT` (`QT` is potentially qualified, e.g. `const`).
+    - `T.deserialize!!T()` should accept one runtime non-default
+       parameter, of the type returned by `T.serialize`.
+
+    The default implementation is the `VibeJSONSerializer`.
+    Other utilities in this module are used to facilitate development
+    and diagnostics.
+
+    Author:         Mathias 'Geod24' Lang
+    License:        MIT (See LICENSE.txt)
+    Copyright:      Copyright (c) 2020 Mathias Lang. All rights reserved.
+
+*******************************************************************************/
+
+module geod24.Serialization;
+
+import std.traits;
+
+/*******************************************************************************
+
+    Serialize arbitrary data to / from JSON
+
+    This is the default serializer used by LocalRest, when none is provided.
+    It is a template so that LocalRest does not import Vibe.d if not used.
+
+*******************************************************************************/
+
+public template VibeJSONSerializer ()
+{
+    import vibe.data.json;
+
+    public string serialize (T) (auto ref T value) @safe
+    {
+        return serializeToJsonString(value);
+    }
+
+    public QT deserialize (QT) (string data) @safe
+    {
+        return deserializeJson!(QT)(data);
+    }
+}
+
+///
+unittest
+{
+    alias S = VibeJSONSerializer!();
+    static assert(!serializerInvalidReason!(S).length, serializerInvalidReason!(S));
+    static assert(is(SerializedT!(S) == string));
+}
+
+/// Utility function to check if a serializer conforms to the requirements
+/// Returns: An error message, or `null` if the serializer is conformant
+public string serializerInvalidReason (alias S) ()
+{
+    // Give codegen a break
+    if (!__ctfe) return null;
+
+    static if (!is(typeof(() { alias X = S.serialize; })))
+        return "`" ~ S.stringof ~ "` is missing a `serialize` method";
+    else static if (!is(typeof(() { alias X = S.deserialize; })))
+        return "`" ~ S.stringof ~ "` is missing a `deserialize` method";
+
+    /// We need a type to deserialize to test this function
+    /// While we should only test with return types / parameter types,
+    /// pretty much anything that can't (de)serialize an `int` is broken.
+
+    else static if (!is(typeof(() { S.serialize(0); })))
```

Comment:
If `serialize` takes a `ref` this won't work, right? Should we relax that restriction?

Reply:
Hum, good point. Personally I think `serialize` should just take `auto ref` (it has to be a template anyway) if it wants to optimize stuff. Otherwise we might end up in a situation where we inadvertently use lvalues here and it breaks in client code.

Of course all of this will be easier once the rvalueref thing is fixed...

Reply:
Ah it has to be templated. Yes that makes sense, I forgot we're dealing with multiple types, doh.